pkg/edgeview/src/tcp.go

```diff
@@ -0,0 +1,723 @@
+		log.Tracef("tcpConnMap(%d) has no chan %d on server, launch", mid, jmsg.ChanNum)
+		tcpDataChn[mid] <- jmsg
+		return
+	}
+	msg := wsMessage{
+		mtype: mtype,
+		msg:   jmsg.Data,
+	}
+	myChan.msgChan <- msg
+}
+
+func cleanClosedMapEntries(idx int) {
+	tcpMapMutex.Lock()
+	deleted := 0
+	recvlocal := 0
+	recvWss := 0
+	for i, m := range tcpConnM[idx].m {
+		if m.closed && time.Since(m.closeTime).Seconds() > 60 {
+			recvlocal += m.recvLocal
+			recvWss += m.recvWss
+			delete(tcpConnM[idx].m, i)
+			deleted++
+		}
+	}
+	log.Tracef("done with cleanup(%d). deleted %d, exist num %d", idx, deleted, len(tcpConnM[idx].m))
+	tcpMapMutex.Unlock()
+}
+
+func doneTCPtransfer(idx int) {
+	tcpMapMutex.Lock()
+	closed := 0
+	for _, m := range tcpConnM[idx].m {
+		if !isClosed(m.done) {
+			close(m.done)
+			closed++
+		}
+	}
+	tcpMapMutex.Unlock()
+	log.Noticef("doneTCPtransfer(%d) closed %d threads", idx, closed)
+}
+
+func (r tcpConnRWMap) Get(ch int) (tcpconn, bool) {
+	tcpMapMutex.Lock()
+	t, ok := r.m[ch]
+	if !ok {
+		tcpMapMutex.Unlock()
+		return tcpconn{}, ok
+	}
+	tcpMapMutex.Unlock()
+	return t, ok
+}
+
+func (r tcpConnRWMap) RecvWssInc(ch int) tcpconn {
+	tcpMapMutex.Lock()
+	m := r.m[ch]
+	m.recvWss++
+	r.m[ch] = m
+	tcpMapMutex.Unlock()
+	return m
+}
+
+func (r tcpConnRWMap) RecvLocalInc(ch int) tcpconn {
+	tcpMapMutex.Lock()
+	m := r.m[ch]
+	m.recvLocal++
+	r.m[ch] = m
+	tcpMapMutex.Unlock()
+	return m
+}
+
+func (r tcpConnRWMap) CloseChan(ch int) {
+	tcpMapMutex.Lock()
+	m := r.m[ch]
+	m.closed = true
+	m.closeTime = time.Now()
+	r.m[ch] = m
+	tcpMapMutex.Unlock()
+}
+
+func (r tcpConnRWMap) PendingChan(ch int) {
+	tcpMapMutex.Lock()
+	m := r.m[ch]
+	m.pending = true
+	r.m[ch] = m
+	tcpMapMutex.Unlock()
+}
+
+func (r tcpConnRWMap) AssignConn(ch int, m tcpconn) chan wsMessage {
+	tcpMapMutex.Lock()
+	r.m[ch] = m
+	tcpMapMutex.Unlock()
+	return m.msgChan
+}
+
+func tcpRecvTimeUpdate() {
+	tcpTimeMutex.Lock()
+	tcpServerRecvTime = time.Now()
+	tcpTimeMutex.Unlock()
+}
+
+func tcpRecvTimeCheckExpire() bool {
+	var expired bool
+	tcpTimeMutex.Lock()
+	if time.Since(tcpServerRecvTime).Seconds() > 600 {
+		expired = true
+	}
+	tcpTimeMutex.Unlock()
+	return expired
+}
+
+func tcpClientSendDone() {
+	if !isTCPClient {
+		wssWrMutex.Lock()
+		sendCloseToWss()
+		wssWrMutex.Unlock()
+		return
+	}
+	wssWrMutex.Lock()
+	// send to server first, then to dispatcher to close
+	_ = addEnvelopeAndWriteWss([]byte(tcpDONEMessage), true)
+	sendCloseToWss()
+	wssWrMutex.Unlock()
+}
+
+func getProxyOpt(opt string) (bool, string) {
+	var proxyDNSIP string
+	var hasProxy bool
+	if strings.HasPrefix(opt, "proxy") {
+		hasProxy = true
+		if strings.Contains(opt, "proxy@") {
+			strs := strings.Split(opt, "proxy@")
+			if len(strs) == 2 {
+				proxyDNSIP = strs[1]
+			}
+		}
+	}
+	return hasProxy, proxyDNSIP
+}
+
+func processTCPcmd(opt string, remotePorts map[int]int) (bool, int, map[int]int) {
+	tcpopts := strings.SplitN(opt, "tcp/", 2)
+	tcpparam := tcpopts[1]
```

Comment:
What syntax do we need to support for tcpparam? Is it a well-known set of IP + port-related synax which we can check for validity.

Reply:
Yes, it can be checked at the top level, at runNetwork() loop for all the options strings. will do.